pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -3942,9 +3942,13 @@ public void publishTxnMessage(TxnID txnID, ByteBuf headersAndPayload, PublishCon
     @Override
     public CompletableFuture<Void> endTxn(TxnID txnID, int txnAction, long lowWaterMark) {
         if (TxnAction.COMMIT_VALUE == txnAction) {
-            return transactionBuffer.commitTxn(txnID, lowWaterMark);
+            return transactionBuffer.commitTxn(txnID, lowWaterMark).thenRun(() -> {
+                lastDataMessagePublishedTimestamp = Clock.systemUTC().millis();
```

Comment:
```
    // Record the last time a data message (ie: not an internal Pulsar marker) is published on the topic
    @Getter
    private volatile long lastDataMessagePublishedTimestamp = 0;
```

For ending transactions, we only put a marker message in the topic. From the description of this field, it shouldn't be updated for internal markers.

Could you please provide more information of which behavior is not expected from the user's perspective?

Reply:
`lastDataMessagePublishedTimestamp` is used to determined whether there are new messages available for reading by `ReplicatedSubscription` feature **only**. So we should update the `lastDataMessagePublishedTimestamp` when the max read position move forward.

Reply:
Oh I see. So, the committed transaction messages will not be replicated to the remote cluster without this fix?

If yes,

I think we should change the PR title to "[fix][broker] fix the transaction messages cannot be replicated"
And, a test for replicating transaction messages should be added to make sure we can fix the issue end to end

Reply:
And the correct fix should be only update the `LastDataMessagePublishedTimestamp` while the max read position is updated. The commit and abort operation can happen without the max read position change.

Reply:
No, we can still trigger the replication by another bug. refer to anther pr:
https://github.com/apache/pulsar/pull/22461

Maybe current test code is enough, this pr is to correct the field `LastDataMessagePublishedTimestamp`.

Reply:
> Oh I see. So, the committed transaction messages will not be replicated to the remote cluster without this fix?
> 
> If yes,
> 
> I think we should change the PR title to "[fix][broker] fix the transaction messages cannot be replicated" And, a test for replicating transaction messages should be added to make sure we can fix the issue end to end

Yes, we will always skip taking snapshot.

Reply:
> Yes, we will always skip taking snapshot. But `ReplicatedSubscription` feature is not used to replicate messages, but replicate subscription state.

@thetumbled I think this could be reflected in the title of the PR. "[fix][broker] fix replicated subscriptions for transactional messages" etc.

Reply:
title changed.

Reply:
@thetumbled The high-level idea in the test looks fine, but it would be better to have it in a completely separate test class since there could be a need to enable transactions etc.. We tend to add too many unrelated tests in single test class and that causes problems. There's no need to have a 1-to-1 mapping between a production code class and a test class. That's leading to some problems in our test design. The test class name could be "TransactionalReplicateSubscriptionTest". (btw. the existing test class name "ReplicatorSubscriptionTest" has a typo).

Reply:
> Yes, it does. But i think that it is a good tradeoff that we allow wrong update, because LastDataMessagePublishedTimestamp is used to skip creating unnecessary subscription replication snapshot.
Updating incorrectly just creating an unnecessary snapshot.
On the contrary, if we chase for strictly correctess, we have to take lots of effort to determine whether the max read position move forward or not.
But we can't allow missing update, which will skip creating necessary subscription replication snapshot.

We already have `updateMaxReadPosition(TxnID txnID)` method in TopicTransactionBuffer.java. And it checked if the maxReadPosition is changed or not. So we can just allow the topic to register a callback to the TopicTransactionBuffer.java

```java
    void updateMaxReadPosition(TxnID txnID) {
        PositionImpl preMaxReadPosition = this.maxReadPosition;
        ongoingTxns.remove(txnID);
        if (!ongoingTxns.isEmpty()) {
            PositionImpl position = ongoingTxns.get(ongoingTxns.firstKey());
            maxReadPosition = ((ManagedLedgerImpl) topic.getManagedLedger()).getPreviousPosition(position);
        } else {
            maxReadPosition = (PositionImpl) topic.getManagedLedger().getLastConfirmedEntry();
        }
        if (preMaxReadPosition.compareTo(this.maxReadPosition) != 0) {
            this.changeMaxReadPositionAndAddAbortTimes.getAndIncrement();
           // new added
            callback.onMaxReadPostionUpdated(this.maxReadPosition);
        }
    }
```

Keep the code easy to understand is important, I mean if someone new to this part to understand why we should update the `LastDataMessagePublishedTimestamp` when adding commit and abort markers to the topic and the description of `LastDataMessagePublishedTimestamp` said the internal marker shouldn't update it.

Reply:
Thanks for reply, i have refactor the implementation, PTAL.